File: python/semantic_kernel/template_engine/code_tokenizer.py
# Copyright (c) Microsoft. All rights reserved.

import logging

from semantic_kernel.exceptions import CodeBlockSyntaxError
from semantic_kernel.template_engine.blocks.block import Block
from semantic_kernel.template_engine.blocks.block_types import BlockTypes
from semantic_kernel.template_engine.blocks.function_id_block import FunctionIdBlock
from semantic_kernel.template_engine.blocks.named_arg_block import NamedArgBlock
from semantic_kernel.template_engine.blocks.symbols import Symbols
from semantic_kernel.template_engine.blocks.val_block import ValBlock
from semantic_kernel.template_engine.blocks.var_block import VarBlock

logger: logging.Logger = logging.getLogger(__name__)


# BNF parsed by CodeTokenizer:
# [template]       ::= "" | [variable] " " [template]
#                         | [value] " " [template]
#                         | [function-call] " " [template]
# [variable]       ::= "$" [valid-name]
# [value]          ::= "'" [text] "'" | '"' [text] '"'
# [function-call]  ::= [function-id] | [function-id] [parameter]
# [parameter]      ::= [variable] | [value]
class CodeTokenizer:
    """Tokenize the code text into blocks."""

    @staticmethod
    def tokenize(text: str) -> list[Block]:
        """Tokenize the code text into blocks."""
        # Remove spaces, which are ignored anyway
        text = text.strip() if text else ""
        # Render None/empty to []
        if not text:
            return []
        # 1 char only edge case, var and val blocks are invalid with one char, so it must be a function id block
        if len(text) == 1:
            return [FunctionIdBlock(content=text)]

        # Track what type of token we're reading
        current_token_type = None

        # Track the content of the current token
        current_token_content: list[str] = []

        # Other state we need to track
        text_value_delimiter = None
        space_separator_found = False
        skip_next_char = False
        next_char = ""
        blocks: list[Block] = []

        for index, current_char in enumerate(text[:-1]):
            next_char = text[index + 1]

            if skip_next_char:
                skip_next_char = False
                continue

            # First char is easy
            if index == 0:
                if current_char == Symbols.VAR_PREFIX:
                    current_token_type = BlockTypes.VARIABLE
                elif current_char in (Symbols.DBL_QUOTE, Symbols.SGL_QUOTE):
                    current_token_type = BlockTypes.VALUE
                    text_value_delimiter = current_char
                else:
                    current_token_type = BlockTypes.FUNCTION_ID

                current_token_content.append(current_char)
                continue

            # While reading values between quotes
            if current_token_type in (BlockTypes.VALUE, BlockTypes.NAMED_ARG):
                # If the current char is escaping the next special char we:
                #  - skip the current char (escape char)
                #  - add the next char (special char)
                #  - jump to the one after (to handle "\\" properly)
                if current_char == Symbols.ESCAPE_CHAR and next_char in (
                    Symbols.DBL_QUOTE,
                    Symbols.SGL_QUOTE,
                    Symbols.ESCAPE_CHAR,
                ):
                    current_token_content.append(next_char)
                    skip_next_char = True
                    continue

                current_token_content.append(current_char)

                # When we reach the end of the value, we add the block
                if current_char == text_value_delimiter:
                    blocks.append(ValBlock(content="".join(current_token_content)))
                    current_token_content.clear()
                    current_token_type = None
                    space_separator_found = False

                continue

            # If we're not between quotes, a space signals the end of the current token
            # Note: there might be multiple consecutive spaces
            if current_char in (
                Symbols.SPACE,
                Symbols.NEW_LINE,
                Symbols.CARRIAGE_RETURN,
                Symbols.TAB,
            ):
                if current_token_type == BlockTypes.VARIABLE:
                    blocks.append(VarBlock(content="".join(current_token_content)))
                    current_token_content.clear()
                elif current_token_type == BlockTypes.FUNCTION_ID:
                    if Symbols.NAMED_ARG_BLOCK_SEPARATOR.value in current_token_content:
                        blocks.append(NamedArgBlock(content="".join(current_token_content)))
                    else:
                        blocks.append(FunctionIdBlock(content="".join(current_token_content)))
                    current_token_content.clear()

                space_separator_found = True
                current_token_type = None

                continue

            # If we're not inside a quoted value, and we're not processing a space
            current_token_content.append(current_char)

            if current_token_type is None:
                if not space_separator_found:
                    raise CodeBlockSyntaxError("Tokens must be separated by one space least")

                if current_char in (Symbols.DBL_QUOTE, Symbols.SGL_QUOTE):
                    # A quoted value starts here
                    current_token_type = BlockTypes.VALUE
                    text_value_delimiter = current_char
                elif current_char == Symbols.VAR_PREFIX:
                    # A variable starts here
                    current_token_type = BlockTypes.VARIABLE
                else:
                    # A function id starts here
                    current_token_type = BlockTypes.FUNCTION_ID

        # end of main for loop

        # Capture last token
        current_token_content.append(next_char)

        if current_token_type == BlockTypes.VALUE:
            blocks.append(ValBlock(content="".join(current_token_content)))
        elif current_token_type == BlockTypes.VARIABLE:
            blocks.append(VarBlock(content="".join(current_token_content)))
        elif current_token_type == BlockTypes.FUNCTION_ID:
            if Symbols.NAMED_ARG_BLOCK_SEPARATOR.value in current_token_content:
                blocks.append(NamedArgBlock(content="".join(current_token_content)))
            else:
                blocks.append(FunctionIdBlock(content="".join(current_token_content)))
        else:
            raise CodeBlockSyntaxError("Tokens must be separated by one space least")

        return blocks

File: nodejs/semantic-kernel/template-engine/code-tokenizer.ts
import { CodeBlockSyntaxError } from '../exceptions/template-engine-exceptions'
import { Block } from './blocks/block'
import { BlockTypes } from './blocks/block-types'
import { FunctionIdBlock } from './blocks/function-id-block'
import { NamedArgBlock } from './blocks/named-arg-block'
import { Symbols } from './blocks/symbols'
import { ValBlock } from './blocks/val-block'
import { VarBlock } from './blocks/var-block'

/**
 * Tokenize the code text into blocks.
 *
 * BNF parsed by CodeTokenizer:
 * [template]       ::= "" | [variable] " " [template]
 *                         | [value] " " [template]
 *                         | [function-call] " " [template]
 * [variable]       ::= "$" [valid-name]
 * [value]          ::= "'" [text] "'" | '"' [text] '"'
 * [function-call]  ::= [function-id] | [function-id] [parameter]
 * [parameter]      ::= [variable] | [value]
 */
export class CodeTokenizer {
  /**
   * Tokenize the code text into blocks.
   * @param text - The code text to tokenize.
   * @returns An array of blocks.
   * @throws CodeBlockSyntaxError if the syntax is invalid.
   */
  public static tokenize(text: string): Block[] {
    // Remove spaces, which are ignored anyway
    text = text ? text.trim() : ''

    // Render None/empty to []
    if (!text) {
      return []
    }

    // 1 char only edge case, var and val blocks are invalid with one char, so it must be a function id block
    if (text.length === 1) {
      return [new FunctionIdBlock({ content: text })]
    }

    // Track what type of token we're reading
    let currentTokenType: BlockTypes | null = null

    // Track the content of the current token
    const currentTokenContent: string[] = []

    // Other state we need to track
    let textValueDelimiter: string | null = null
    let spaceSeparatorFound = false
    let skipNextChar = false
    let nextChar = ''
    const blocks: Block[] = []

    for (let index = 0; index < text.length - 1; index++) {
      const currentChar = text[index]
      nextChar = text[index + 1]

      if (skipNextChar) {
        skipNextChar = false
        continue
      }

      // First char is easy
      if (index === 0) {
        if (currentChar === Symbols.VAR_PREFIX) {
          currentTokenType = BlockTypes.VARIABLE
        } else if (currentChar === Symbols.DBL_QUOTE || currentChar === Symbols.SGL_QUOTE) {
          currentTokenType = BlockTypes.VALUE
          textValueDelimiter = currentChar
        } else {
          currentTokenType = BlockTypes.FUNCTION_ID
        }

        currentTokenContent.push(currentChar)
        continue
      }

      // While reading values between quotes
      if (currentTokenType === BlockTypes.VALUE) {
        // If the current char is escaping the next special char we:
        //  - skip the current char (escape char)
        //  - add the next char (special char)
        //  - jump to the one after (to handle "\\" properly)
        if (
          currentChar === Symbols.ESCAPE_CHAR &&
          (nextChar === Symbols.DBL_QUOTE || nextChar === Symbols.SGL_QUOTE || nextChar === Symbols.ESCAPE_CHAR)
        ) {
          currentTokenContent.push(nextChar)
          skipNextChar = true
          continue
        }

        currentTokenContent.push(currentChar)

        // When we reach the end of the value, we add the block
        if (currentChar === textValueDelimiter) {
          blocks.push(new ValBlock({ content: currentTokenContent.join('') }))
          currentTokenContent.length = 0
          currentTokenType = null
          spaceSeparatorFound = false
        }

        continue
      }

      // If we're not between quotes, a space signals the end of the current token
      // Note: there might be multiple consecutive spaces
      if (
        currentChar === Symbols.SPACE ||
        currentChar === Symbols.NEW_LINE ||
        currentChar === Symbols.CARRIAGE_RETURN ||
        currentChar === Symbols.TAB
      ) {
        if (currentTokenType === BlockTypes.VARIABLE) {
          blocks.push(new VarBlock({ content: currentTokenContent.join('') }))
          currentTokenContent.length = 0
        } else if (currentTokenType === BlockTypes.FUNCTION_ID) {
          if (currentTokenContent.includes(Symbols.NAMED_ARG_BLOCK_SEPARATOR)) {
            blocks.push(new NamedArgBlock({ content: currentTokenContent.join('') }))
          } else {
            blocks.push(new FunctionIdBlock({ content: currentTokenContent.join('') }))
          }
          currentTokenContent.length = 0
        }

        spaceSeparatorFound = true
        currentTokenType = null

        continue
      }

      // If we're not inside a quoted value, and we're not processing a space
      currentTokenContent.push(currentChar)

      if (currentTokenType === null) {
        if (!spaceSeparatorFound) {
          throw new CodeBlockSyntaxError('Tokens must be separated by one space least')
        }

        if (currentChar === Symbols.DBL_QUOTE || currentChar === Symbols.SGL_QUOTE) {
          // A quoted value starts here
          currentTokenType = BlockTypes.VALUE
          textValueDelimiter = currentChar
        } else if (currentChar === Symbols.VAR_PREFIX) {
          // A variable starts here
          currentTokenType = BlockTypes.VARIABLE
        } else {
          // A function id starts here
          currentTokenType = BlockTypes.FUNCTION_ID
        }
      }
    }

    // end of main for loop

    // Capture last token
    currentTokenContent.push(nextChar)

    if (currentTokenType === BlockTypes.VALUE) {
      blocks.push(new ValBlock({ content: currentTokenContent.join('') }))
    } else if (currentTokenType === BlockTypes.VARIABLE) {
      blocks.push(new VarBlock({ content: currentTokenContent.join('') }))
    } else if (currentTokenType === BlockTypes.FUNCTION_ID) {
      if (currentTokenContent.includes(Symbols.NAMED_ARG_BLOCK_SEPARATOR)) {
        blocks.push(new NamedArgBlock({ content: currentTokenContent.join('') }))
      } else {
        blocks.push(new FunctionIdBlock({ content: currentTokenContent.join('') }))
      }
    } else {
      throw new CodeBlockSyntaxError('Tokens must be separated by one space least')
    }

    return blocks
  }
}
